oak_attestation_verification/src/verifier.rs

```diff
@@ -121,43 +122,166 @@ fn verify_dice_chain(evidence: &Evidence) -> anyhow::Result<()> {
 }
 
 fn verify_oak_restricted_kernel(
-    _evidence: &Evidence,
-    _endorsements: &OakRestrictedKernelEndorsements,
-    _reference_values: &OakRestrictedKernelReferenceValues,
+    now_utc_millis: i64,
+    endorsements: &OakRestrictedKernelEndorsements,
+    reference_values: &OakRestrictedKernelReferenceValues,
 ) -> anyhow::Result<()> {
-    anyhow::bail!("Needs implementation")
+    // Bind references to layers to variables, for both endorsements and
+    // reference values. The presence of all fields is necessary for
+    // verification success - there are no exceptions.
+    let e0 = endorsements
+        .root_layer
+        .as_ref()
+        .ok_or_else(|| anyhow::anyhow!("no root layer endorsements"))?;
+    let e1 = endorsements
+        .kernel_layer
+        .as_ref()
+        .ok_or_else(|| anyhow::anyhow!("no kernel layer endorsements"))?;
+    let e2 = endorsements
+        .application_layer
+        .as_ref()
+        .ok_or_else(|| anyhow::anyhow!("no application layer endorsements"))?;
+
+    let r0 = reference_values
+        .root_layer
+        .as_ref()
+        .ok_or_else(|| anyhow::anyhow!("no root layer reference values"))?;
+    let r1 = reference_values
+        .kernel_layer
+        .as_ref()
+        .ok_or_else(|| anyhow::anyhow!("no root layer reference values"))?;
+    let r2 = reference_values
+        .application_layer
+        .as_ref()
+        .ok_or_else(|| anyhow::anyhow!("no root layer reference values"))?;
+
+    // Verify Transparent Release endorsements in root layer.
+    verify_root_layer_endorsements(now_utc_millis, e0, r0)?;
+
+    // Verify Transparent Release endorsements in kernel layer.
+    verify_kernel_layer_endorsements(now_utc_millis, e1, r1)?;
+
+    // Verify Transparent Release endorsements in application layer.
+    let e20 = e2
+        .binary
+        .as_ref()
+        .ok_or_else(|| anyhow::anyhow!("no binary endorsement"))?;
+
+    let r20 = r2
+        .binary
+        .as_ref()
+        .ok_or_else(|| anyhow::anyhow!("no binary reference value"))?;
+
+    verify_transparent_release_endorsement(now_utc_millis, e20, r20)?;
+
+    let e21 = e2
+        .configuration
+        .as_ref()
+        .ok_or_else(|| anyhow::anyhow!("no configuration endorsement"))?;
+
+    let r21 = r2
+        .configuration
+        .as_ref()
+        .ok_or_else(|| anyhow::anyhow!("no configuration reference value"))?;
+
+    verify_transparent_release_endorsement(now_utc_millis, e21, r21)
 }
 
 fn verify_oak_containers(
     now_utc_millis: i64,
-    _evidence: &Evidence,
     endorsements: &OakContainersEndorsements,
     reference_values: &OakContainersReferenceValues,
 ) -> anyhow::Result<()> {
-    let ends_layer = endorsements
+    // Bind references to layers to variables.
+    let e0 = endorsements
```

Comment:
Would it be possible to use more descriptive variable names, e.g., root_layer_endorsements/ref_values? Or maybe even use the fields directly, especially if you create separate functions for layers (see my next suggestion)?

Reply:
+1